File: 01-Programming-Languages/R/Algorithms/Recursion/tower_of_hanoi.R
# Problem: Tower of Hanoi
# Description: Move n disks from source to destination using an auxiliary peg.
#
# Approach:
# - Move n-1 disks from source to auxiliary
# - Move nth disk from source to destination
# - Move n-1 disks from auxiliary to destination
#
# Time Complexity: O(2^n)
# Space Complexity: O(n) (recursion stack)
#
# Example:
# Input: n = 3, source = "A", destination = "C", auxiliary = "B"
# Output: Moves to transfer all disks

tower_of_hanoi <- function(n, source, destination, auxiliary) {
  if (n == 1) {
    cat("Move disk 1 from", source, "to", destination, "\n")
    return()
  }
  tower_of_hanoi(n - 1, source, auxiliary, destination)
  cat("Move disk", n, "from", source, "to", destination, "\n")
  tower_of_hanoi(n - 1, auxiliary, destination, source)
}

cat("Tower of Hanoi - 3 disks:\n")
tower_of_hanoi(3, "A", "C", "B")

File: 01-Programming-Languages/Dart/Algorithms/Recursion/tower_of_hanoi.dart
/*
Problem: Tower of Hanoi
Description: Solve the Tower of Hanoi puzzle. Move n disks from source
           peg to destination peg using an auxiliary peg.

Approach:
- Recursively move n-1 disks from source to auxiliary
- Move the nth disk from source to destination
- Recursively move n-1 disks from auxiliary to destination

Time Complexity: O(2^n)
Space Complexity: O(n) for recursion stack

Example:
Input: n = 3, source = "A", destination = "C", auxiliary = "B"
Output: Move disk 1 from A to C, ...
*/

int moveCount = 0;

void towerOfHanoi(int n, String source, String destination, String auxiliary) {
  if (n == 1) {
    moveCount++;
    print('Move disk 1 from $source to $destination');
    return;
  }
  towerOfHanoi(n - 1, source, auxiliary, destination);
  moveCount++;
  print('Move disk $n from $source to $destination');
  towerOfHanoi(n - 1, auxiliary, destination, source);
}

void main() {
  moveCount = 0;
  towerOfHanoi(3, "A", "C", "B");
  print('Total moves: $moveCount');
}
